File: java/FileIndex.scala
package CQLConnect

import java.util.Date
import scala.language.implicitConversions

import com.datastax.driver.core.{ Row, LocalDate }
import com.weather.scalacass._
import com.weather.scalacass.syntax._

import CQLConnect.Models.{ DistMVFileIndexModel, extract_dist_mvfileindex_values }


object FileIndex {

  def query_dist_file_index( session:ScalaSession ) : Iterator[Row] = {
    val keyspace = session.keyspace

    val query =
      s"""SELECT
         |  line,
         |  analysed,
         |  date,
         |  file_name,
         |  state
         |FROM $keyspace.mv_file_index""".stripMargin

    session.rawSelect(query)

  }

  def get_dist_file_index_row( session:ScalaSession ) : Iterator[Row] = {
    // Query cassandra to get the row, converting R values to Cassandra mappable classes
    query_dist_file_index(session)
  }

  def iterator_exhausted( df:Iterator[Row] ) : Boolean = {
    df.hasNext
  }

  def get_dist_file_index( df:Iterator[Row] ) : Array[Any] = {
    // If there is no row, return an empty array
    if(df.isEmpty) { return(Array():Array[Any]) }
    // If a row exists, puts the row into a model
    val row: Row = df.next()
    val model: DistMVFileIndexModel = row.as[DistMVFileIndexModel]
    // Return the extracted values from the model in a list
    extract_dist_mvfileindex_values(model)
  }

}

File: R/file_index.R
#' File Index module
#'
#' Build a CQL query for a precompiled model
#' Pull the retuned objects back into R
#'
#' @rdname DistMvFileIndex
#'
#' @param sc Spark context used for the scala context
#' @param session The cassandra session
NULL

#' Dispatch the Distinct File Index query
#'
#' @export
cql_dist_file_index_query <- function(sc, session) {
  sparklyr::invoke_static(sc, "CQLConnect.FileIndex", "query_dist_file_index", session)
}

#' Extract file_index row
#'
#' @include utils.R
#' @export
cql_get_file_index_row <- function(sc, session) {
  sparklyr::invoke_static(sc, "CQLConnect.FileIndex", "get_dist_file_index_row", session)
}

#' Exhaust file_index row iterator to get whole file_index
#'
#' @include utils.R
#' @export
cql_iterate <- function(sc, iterator) {
  sparklyr::invoke_static(sc, "CQLConnect.FileIndex", "get_dist_file_index", iterator)
}

#' Exhaust file_index row iterator to get whole file_index
#'
#' @include utils.R
#' @export
cql_hasnext <- function(sc, iterator) {
  sparklyr::invoke_static(sc, "CQLConnect.FileIndex", "iterator_exhausted", iterator)
}

#' Exhaust file_index row iterator to get whole file_index
#'
#' @import data.table
#' @include utils.R
#' @export
cql_get_file_index <- function(sc, session) {

  # Table count to preallocate memory for the iterator
  n = cql_count(sc, session, "mv_file_index")

  # Iterator of a table row
  iterator = cql_get_file_index_row(sc, session)

  # Exhaust the iterator through a map
  dtl = lapply(seq_along(1:n), function(i) {
    sparklyr::invoke_static(sc, "CQLConnect.FileIndex", "get_dist_file_index", iterator)
  })

  # Transpose the nested list and set as a data.table
  dt = split(unlist(dtl, use.names = FALSE), c("line", "analysed", "date", "file_name", "state"))
  data.table::setDT(dt)

  # Some casting which could be avoided if not for unlist
  dt[, analysed := as.logical(analysed)]
  dt[, line := as.integer(line)]
  dt[, date := as.Date(as.numeric(date), origin = "1970-01-01")]

  # Call [.data.table so function returns an object
  dt[]
}
